README.md
# Voting-Optimization

## Introduction
We attempt to solve multi-objective optimization problems, specifically Student Project Allocation, by extending the literature of social choice theory.
This repository constains all the code for the dissertation.

## Setup

To use the code in this directory you first need to install the required packages using requirements.txt.
If you wish to simply reproduce our results then you can choose to stop here and move onto the directory structure.


Please note to use the accumulation script(`project_allocation_accumulator.py`) you will need to install MiniZinc and Gurobi seperately on your machine.

You can use the following installation guides:
https://www.gurobi.com/documentation/9.5/quickstart_mac/software_installation_guid.html
https://www.minizinc.org/doc-2.2.3/en/installation.html

For the networkflow approach for solving the problem you may refer to the jar executable in the directory. Please note that you will need to have java installed for this and I'm not the author of this software.


## Directory Structure
```
Instances - i.e 2015, 2020, 2021, random_1, similar_1, these are the 5 instances we use in the dissertation
    data - this directory constrains all the solutions gathered from the accumulation script
    instance_name_jar.txt - this file is the input file for the jar file that constains all the implementations of the network flow algorithms
    instance_name.dzn - this in the MiniZinc datafile for the instance
    project_allocation_instance_name.py - this is the python file that generates the MiniZinc datafile.
borda_runner.py - this is the main file that runs the borda count algorithm, you just run it like a normal python file and it will ask you for an instance name, you can provide it with any of the 5 mentioned above. i.e 2015, 2020, 2021, random_1, similar_1. It returns the ids of the winning solution

copeland_runner.py - works the same way as borda runner.
schulze_runner.py - works the same way as borda runner.
winner_analysis.py - you run the python file, it will ask for an instance name, you can provide it with any of the 5 mentioned above. It then asks for an ID of the solution, you can use any that you get from the voting algorithms. It prints out the profile and cost of each solution.
```


2021/changes.txt
n = 190;
p = 394;
l = 181;
[111, 342, 49, 57, 129, 110, 338, 131],
[342, 257, 11, 218, 364, 48, 174, 363],
[330, 256, 341, 285, 210, 362, 313, 304],
[55, 333, 162, 72, 330, 64, 165, 294],
[259, 341, 40, 256, 340, 235, 10, 55],
[257, 251, 341, 59, 304, 333, 111, 10],
[48, 57, 362, 22, 160, 335, 116, 129],
[263, 264, 43, 44, 55, 360, 333, 4],
[216, 224, 225, 179, 215, 107, 291, 105],
[350, 358, 278, 196, 87, 85, 270, 201],
[273, 357, 152, 281, 346, 196, 306, 110],
[196, 276, 203, 279, 267, 238, 258, 307],
[251, 136, 258, 304, 285, 43, 236, 135],
[328, 162, 333, 18, 155, 43, 167, 251],
[257, 190, 135, 350, 151, 123, 60, 274],
[105, 217, 13, 168, 257, 111, 74, 159],
[251, 260, 59, 55, 44, 193, 209, 14],
[333, 149, 244, 184, 44, 105, 155, 162],
[276, 154, 261, 257, 135, 136, 241, 236],
[285, 255, 251, 346, 136, 278, 198, 235],
[57, 48, 291, 51, 167, 116, 158, 281],
[57, 344, 247, 48, 46, 297, 279, 329],
[251, 59, 49, 168, 41, 158, 55, 141],
[4, 1, 322, 317, 232, 242, 340, 263],
[276, 278, 356, 355, 267, 103, 298, 196],
[119, 171, 284, 294, 145, 68, 268, 332],
[346, 257, 48, 44, 340, 105, 254, 43],
[322, 242, 394, 339, 193, 190, 207, 304],
[10, 304, 360, 346, 150, 44, 8, 328],
[168, 18, 162, 319, 19, 86, 153, 333],
[219, 218, 174, 104, 11, 230, 175, 336],
[326, 29, 297, 212, 100, 105, 185, 18],
[341, 111, 43, 48, 154, 20, 70, 250],
[172, 307, 44, 297, 251, 63, 361, 285],
[154, 55, 59, 347, 346, 111, 239, 236],
[105, 43, 38, 174, 216, 230, 326, 10],
[341, 360, 261, 150, 285, 259, 332],
[4, 328, 5, 27, 333, 181, 155, 319],
[236, 241, 304, 308, 10, 251, 285, 257],
[44, 251, 99, 358, 37, 353, 85, 98],
[333, 120, 331, 156, 216, 296, 107, 25],
[140, 362, 334, 59, 285, 111, 48, 43],
[151, 150, 49, 314, 162, 241, 247, 338],
[111, 251, 48, 285, 341, 55, 43, 304],
[45, 216, 224, 105, 299, 225, 174, 18],
[39, 296, 225, 220, 174, 298, 70, 202],
[111, 341, 332, 353, 336, 55, 234, 285],
[337, 338, 55, 100, 56, 99, 226, 174],
[55, 350, 267, 77, 70, 310, 232, 210],
[236, 235, 360, 341, 281, 285, 353],
[254 264, 77, 341, 304, 43, 44, 285],
[163, 240, 16, 256, 37, 307, 321, 346],
[165, 164, 141, 341, 58, 330, 37, 29],
[161, 242, 4, 376, 2, 316, 313, 195],
[297, 216, 179, 224, 20, 175, 167, 41],
[266, 286, 302, 84, 176, 49, 39, 297],
[336, 380, 377, 312, 320, 78, 20, 246],
[146, 178, 38, 110, 342, 105, 247],
[368, 366, 203, 285, 199, 207, 84, 266],
[261, 239, 154, 257, 177, 232, 48, 276],
[321, 40, 257, 137, 172, 192, 268, 250],
[174, 175, 216, 213, 96, 226, 105, 340],
[319, 328, 333, 366, 168, 368, 288, 289],
[40, 341, 332, 44, 328, 111, 105, 299],
[154, 105, 172, 43, 48, 216, 158, 294],
[10, 120, 176, 212, 165, 137, 353, 179],
[53, 360, 111, 231, 341, 322, 46, 326],
[43, 39, 315, 168, 321, 332, 285],
[44, 137, 333, 291, 145, 107, 289, 211],
[29, 162, 333, 139, 40, 332, 296, 175],
[168, 307, 167, 333, 304, 44, 86, 240],
[326, 135, 172, 100, 4, 217, 76, 154],
[158, 41, 46, 250, 255, 241, 313, 308],
[196, 124, 85, 203, 87, 355, 274, 94],
[257, 359, 261, 341, 320, 306, 111, 190],
[45, 297, 111, 105, 236, 48, 308, 85],
[314, 160, 315, 43, 144, 297, 40, 230],
[20, 43, 39, 116, 158, 291, 160, 293],
[87, 89, 359, 44, 360, 150, 167, 304],
[2, 322, 5, 190, 242, 195, 263, 15],
[143, 333, 55, 259, 110, 48, 308, 341],
[55, 346, 53, 257, 259, 48, 236, 320],
[355, 87, 135, 357, 88, 136, 158, 104],
[371, 22, 159, 80, 359, 65, 133, 23],
[87, 89, 111, 353, 341, 231, 365, 257],
[104, 211, 163, 328, 160, 326, 149, 98],
[353, 86, 351, 87, 40, 163, 167, 304],
[285, 236, 59, 60, 37, 16, 308, 258],
[207, 198, 266, 201, 355, 254, 48, 293],
[55, 49, 60, 41, 43, 20, 16, 100],
[258, 308, 205, 136, 285, 362, 172, 86],
[272, 210, 327, 230, 333, 4, 70, 11],
[48, 255, 346, 309, 343, 86, 44, 43],
[291, 49, 297, 111, 357, 158, 240, 340],
[364, 331, 292, 391, 172, 251, 334, 69],
[167, 333, 111, 135, 18, 299, 65, 364],
[308, 285, 232, 320, 236, 341, 56, 359],
[43, 49, 44, 236, 232, 253, 265, 105],
[350, 210, 55, 341, 382, 251, 232],
[49, 39, 48, 13, 337, 43, 84, 232],
[199, 86, 280, 285, 239, 355, 209, 360],
[10, 304, 241, 285, 308, 258, 353, 281],
[149, 87, 84, 271, 352, 196, 37, 199],
[154, 70, 258, 49, 308, 292, 316, 196],
[353, 241, 236, 257, 261, 46, 284, 154],
[105, 48, 44, 292, 237, 257, 111, 361],
[286, 84, 270, 281, 85, 155, 259],
[265, 37, 74, 320, 359, 341, 59, 293],
[385, 393, 330, 158, 76, 265, 153, 263],
[41, 172, 40, 173, 211, 165, 164, 294],
[135, 136, 196, 199, 87, 86, 304, 354],
[162, 328, 167, 43, 251, 44, 297, 70],
[360, 285, 251, 43, 239, 154, 235, 111],
[84, 276, 352, 269, 85, 357, 150],
[282, 298, 29, 367, 203, 14, 258],
[336, 43, 326, 341, 49, 133, 320, 257],
[322, 273, 341, 353, 3, 175, 382, 264],
[276, 105, 201, 48, 86, 350, 257, 56],
[188, 199, 276, 146, 40, 4, 43, 362],
[388, 20, 49, 353, 236, 262, 44, 281],
[49, 46, 43, 111, 158, 110, 285, 59],
[230, 224, 175, 216, 294, 116, 29],
[331, 29, 171, 220, 64, 155, 65, 141]
[49, 350, 74, 18, 285, 343, 190, 136],
[39, 173, 309, 45, 70, 313, 341, 357],
[150, 44, 300, 173, 298, 37, 195, 253],
[364, 140, 255, 40, 342, 304, 285, 72],
[18, 256, 262, 350, 307, 308, 10, 333],
[276, 196, 111, 87, 10, 355, 44, 294],
[236, 285, 241, 251, 43, 281, 293, 333],
[39, 44, 314, 123, 355, 208, 96],
[49, 233, 232, 342, 360, 264, 263, 208],
[86, 307, 20, 44, 308, 340, 313, 251],
[45, 285, 46, 154, 241, 261, 236, 158],
[48, 307, 346, 43, 251, 49, 232, 297],
[55, 343, 13, 59, 350, 342, 361, 111],
[353, 239, 70, 307, 260, 40, 177, 210],
[195, 308, 304, 251, 4, 8, 43, 193],
[135, 136, 308, 304, 43, 44, 262, 251],
[99, 284, 333, 87, 40, 241, 360, 321],
[59, 55, 255, 276, 394, 153, 275, 200],
[262, 236, 18, 55, 135, 256, 16],
[46, 257, 43, 341, 251, 315, 172, 314]
[360, 46, 172, 105, 295, 156, 333, 143],
[128, 46, 191, 50, 349, 58, 299, 219],
[104, 154, 98, 110, 48, 164, 261, 310],
[259, 231, 44, 308, 298, 147, 106, 346],
[107, 38, 108, 185, 46, 314, 143, 48],
[92, 208, 59, 346, 86, 200, 307, 255],
[270, 159, 357, 48, 111, 154, 346, 46],
[84, 307, 87, 196, 285, 304, 274, 199],

2021/project_allocation2021.py
import numpy as np

projects_without_index = list(range(1,395))
project_supervisors_map = [101, 101, 101, 101, 101, 101, 101, 101, 111, 111, 111, 111, 111, 110, 110, 110, 110, 110, 110, 110, 110, 110, 110, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 103, 103, 103, 103, 103, 103, 103, 103, 103, 103, 103, 105, 105, 105, 105, 1, 1, 1, 1, 1, 1, 1, 1, 1, 53, 53, 53, 53, 53, 53, 53, 53, 53, 53, 158, 158, 158, 158, 158, 158, 158, 158, 158, 158, 158, 158, 158, 51, 51, 51, 51, 51, 51, 51, 51, 51, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 102, 8, 8, 8, 8, 8, 8, 120, 120, 120, 120, 120, 120, 120, 106, 106, 106, 106, 106, 106, 106, 106, 56, 56, 56, 56, 56, 56, 56, 56, 56, 56, 68, 68, 33, 132, 132, 132, 132, 132, 132, 132, 132, 132, 132, 57, 57, 57, 57, 57, 57, 57, 57, 57, 174, 174, 174, 174, 174, 76, 76, 76, 76, 76, 76, 76, 76, 85, 85, 85, 85, 85, 85, 85, 85, 85, 85, 23, 23, 23, 23, 23, 23, 23, 97, 97, 97, 97, 97, 97, 97, 97, 97, 175, 175, 175, 175, 175, 175, 175, 175, 175, 175, 175, 100, 100, 100, 100, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 181, 181, 181, 181, 181, 181, 181, 181, 181, 181, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 28, 49, 178, 178, 153, 153, 153, 153, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 80, 80, 107, 107, 107, 107, 107, 107, 107, 107, 107, 107, 107, 107, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 140, 140, 140, 140, 140, 140, 140, 140, 140, 140, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 34, 34, 70, 70, 70, 70, 70, 70, 70, 70, 70, 70, 155, 155, 155, 155, 142, 142, 104, 104, 104, 104, 104, 92, 54, 54, 54, 54, 54, 125, 125, 125, 125, 125, 125, 125, 125, 125, 125, 18, 18, 18, 18, 18, 18, 18, 18, 18, 108, 171, 171, 171, 171, 171, 161, 161, 161, 161, 117, 117, 117, 117, 117, 117, 117, 117, 117, 117, 168, 168, 168, 168, 168, 168, 43, 43, 43, 43, 43, 43, 43, 43, 43, 43]
assert len(projects_without_index) == len(project_supervisors_map)
project_supervisor_zip = zip(projects_without_index, project_supervisors_map)

supervisors = list(range(1,182))
lecturerMax = [3, 3, 3, 3, 0, 0, 3, 0, 3, 3, 3, 3, 3, 3, 3, 3, 0, 0, 0, 3, 3, 3, 0, 0, 3, 3, 3, 2, 3, 3, 3, 3, 0, 1, 3, 3, 0, 3, 3, 3, 3, 3, 0, 3, 2, 3, 3, 3, 2, 3, 0, 3, 0, 0, 3, 0, 0, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 1, 3, 0, 3, 3, 3, 3, 3, 0, 3, 3, 3, 0, 3, 3, 3, 3, 0, 3, 3, 3, 3, 3, 3, 1, 2, 3, 3, 3, 0, 2, 3, 0, 0, 2, 0, 0, 1, 0, 0, 1, 3, 0, 0, 3, 0, 3, 3, 3, 0, 3, 3, 0, 3, 3, 3, 3, 0, 3, 3, 3, 3, 3, 3, 0, 0, 3, 3, 3, 3, 3, 3, 0, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 1, 3, 1, 3, 3, 0, 3, 3, 0, 3, 3, 3, 3, 3, 3, 1, 3, 3, 1, 3, 3, 0, 0, 3, 3, 2, 3, 3, 0]
assert len(supervisors) == len(lecturerMax)
supervisor_max_zip = zip(supervisors, lecturerMax)

studentProjectRanks = [
    [111, 342, 49, 57, 129, 110, 338, 131],
    [342, 257, 11, 218, 364, 48, 174, 363],
    [330, 256, 341, 285, 210, 362, 313, 304],
    [55, 333, 162, 72, 330, 64, 165, 294],
    [259, 341, 40, 256, 340, 235, 10, 55],
    [257, 251, 341, 59, 304, 333, 111, 10],
    [48, 57, 362, 22, 160, 335, 116, 129],
    [263, 264, 43, 44, 55, 360, 333, 4],
    [216, 224, 225, 179, 215, 107, 291, 105],
    [350, 358, 278, 196, 87, 85, 270, 201],
    [273, 357, 152, 281, 346, 196, 306, 110],
    [196, 276, 203, 279, 267, 238, 258, 307],
    [251, 136, 258, 304, 285, 43, 236, 135],
    [328, 162, 333, 18, 155, 43, 167, 251],
    [257, 190, 135, 350, 151, 123, 60, 274],
    [105, 217, 13, 168, 257, 111, 74, 159],
    [251, 260, 59, 55, 44, 193, 209, 14],
    [333, 149, 244, 184, 44, 105, 155, 162],
    [276, 154, 261, 257, 135, 136, 241, 236],
    [285, 255, 251, 346, 136, 278, 198, 235],
    [57, 48, 291, 51, 167, 116, 158, 281],
    [57, 344, 247, 48, 46, 297, 279, 329],
    [251, 59, 49, 168, 41, 158, 55, 141],
    [4, 1, 322, 317, 232, 242, 340, 263],
    [276, 278, 356, 355, 267, 103, 298, 196],
    [119, 171, 284, 294, 145, 68, 268, 332],
    [346, 257, 48, 44, 340, 105, 254, 43],
    [322, 242, 394, 339, 193, 190, 207, 304],
    [10, 304, 360, 346, 150, 44, 8, 328],
    [168, 18, 162, 319, 19, 86, 153, 333],
    [219, 218, 174, 104, 11, 230, 175, 336],
    [326, 29, 297, 212, 100, 105, 185, 18],
    [341, 111, 43, 48, 154, 20, 70, 250],
    [172, 307, 44, 297, 251, 63, 361, 285],
    [154, 55, 59, 347, 346, 111, 239, 236],
    [105, 43, 38, 174, 216, 230, 326, 10],
    [341, 360, 261, 150, 285, 259, 332],
    [4, 328, 5, 27, 333, 181, 155, 319],
    [236, 241, 304, 308, 10, 251, 285, 257],
    [44, 251, 99, 358, 37, 353, 85, 98],
    [333, 120, 331, 156, 216, 296, 107, 25],
    [140, 362, 334, 59, 285, 111, 48, 43],
    [151, 150, 49, 314, 162, 241, 247, 338],
    [111, 251, 48, 285, 341, 55, 43, 304],
    [45, 216, 224, 105, 299, 225, 174, 18],
    [39, 296, 225, 220, 174, 298, 70, 202],
    [111, 341, 332, 353, 336, 55, 234, 285],
    [337, 338, 55, 100, 56, 99, 226, 174],
    [55, 350, 267, 77, 70, 310, 232, 210],
    [236, 235, 360, 341, 281, 285, 353],
    [254, 264, 77, 341, 304, 43, 44, 285],
    [163, 240, 16, 256, 37, 307, 321, 346],
    [165, 164, 141, 341, 58, 330, 37, 29],
    [161, 242, 4, 376, 2, 316, 313, 195],
    [297, 216, 179, 224, 20, 175, 167, 41],
    [266, 286, 302, 84, 176, 49, 39, 297],
    [336, 380, 377, 312, 320, 78, 20, 246],
    [146, 178, 38, 110, 342, 105, 247],
    [368, 366, 203, 285, 199, 207, 84, 266],
    [261, 239, 154, 257, 177, 232, 48, 276],
    [321, 40, 257, 137, 172, 192, 268, 250],
    [174, 175, 216, 213, 96, 226, 105, 340],
    [319, 328, 333, 366, 168, 368, 288, 289],
    [40, 341, 332, 44, 328, 111, 105, 299],
    [154, 105, 172, 43, 48, 216, 158, 294],
    [10, 120, 176, 212, 165, 137, 353, 179],
    [53, 360, 111, 231, 341, 322, 46, 326],
    [43, 39, 315, 168, 321, 332, 285],
    [44, 137, 333, 291, 145, 107, 289, 211],
    [29, 162, 333, 139, 40, 332, 296, 175],
    [168, 307, 167, 333, 304, 44, 86, 240],
    [326, 135, 172, 100, 4, 217, 76, 154],
    [158, 41, 46, 250, 255, 241, 313, 308],
    [196, 124, 85, 203, 87, 355, 274, 94],
    [257, 359, 261, 341, 320, 306, 111, 190],
    [45, 297, 111, 105, 236, 48, 308, 85],
    [314, 160, 315, 43, 144, 297, 40, 230],
    [20, 43, 39, 116, 158, 291, 160, 293],
    [87, 89, 359, 44, 360, 150, 167, 304],
    [2, 322, 5, 190, 242, 195, 263, 15],
    [143, 333, 55, 259, 110, 48, 308, 341],
    [55, 346, 53, 257, 259, 48, 236, 320],
    [355, 87, 135, 357, 88, 136, 158, 104],
    [371, 22, 159, 80, 359, 65, 133, 23],
    [87, 89, 111, 353, 341, 231, 365, 257],
    [104, 211, 163, 328, 160, 326, 149, 98],
    [353, 86, 351, 87, 40, 163, 167, 304],
    [285, 236, 59, 60, 37, 16, 308, 258],
    [207, 198, 266, 201, 355, 254, 48, 293],
    [55, 49, 60, 41, 43, 20, 16, 100],
    [258, 308, 205, 136, 285, 362, 172, 86],
    [272, 210, 327, 230, 333, 4, 70, 11],
    [48, 255, 346, 309, 343, 86, 44, 43],
    [291, 49, 297, 111, 357, 158, 240, 340],
    [364, 331, 292, 391, 172, 251, 334, 69],
    [167, 333, 111, 135, 18, 299, 65, 364],
    [308, 285, 232, 320, 236, 341, 56, 359],
    [43, 49, 44, 236, 232, 253, 265, 105],
    [350, 210, 55, 341, 382, 251, 232],
    [49, 39, 48, 13, 337, 43, 84, 232],
    [199, 86, 280, 285, 239, 355, 209, 360],
    [10, 304, 241, 285, 308, 258, 353, 281],
    [149, 87, 84, 271, 352, 196, 37, 199],
    [154, 70, 258, 49, 308, 292, 316, 196],
    [353, 241, 236, 257, 261, 46, 284, 154],
    [105, 48, 44, 292, 237, 257, 111, 361],
    [286, 84, 270, 281, 85, 155, 259],
    [265, 37, 74, 320, 359, 341, 59, 293],
    [385, 393, 330, 158, 76, 265, 153, 263],
    [41, 172, 40, 173, 211, 165, 164, 294],
    [135, 136, 196, 199, 87, 86, 304, 354],
    [162, 328, 167, 43, 251, 44, 297, 70],
    [360, 285, 251, 43, 239, 154, 235, 111],
    [84, 276, 352, 269, 85, 357, 150],
    [282, 298, 29, 367, 203, 14, 258],
    [336, 43, 326, 341, 49, 133, 320, 257],
    [322, 273, 341, 353, 3, 175, 382, 264],
    [276, 105, 201, 48, 86, 350, 257, 56],
    [188, 199, 276, 146, 40, 4, 43, 362],
    [388, 20, 49, 353, 236, 262, 44, 281],
    [49, 46, 43, 111, 158, 110, 285, 59],
    [230, 224, 175, 216, 294, 116, 29],
    [331, 29, 171, 220, 64, 155, 65, 141],
    [49, 350, 74, 18, 285, 343, 190, 136],
    [39, 173, 309, 45, 70, 313, 341, 357],
    [150, 44, 300, 173, 298, 37, 195, 253],
    [364, 140, 255, 40, 342, 304, 285, 72],
    [18, 256, 262, 350, 307, 308, 10, 333],
    [276, 196, 111, 87, 10, 355, 44, 294],
    [236, 285, 241, 251, 43, 281, 293, 333],
    [39, 44, 314, 123, 355, 208, 96],
    [49, 233, 232, 342, 360, 264, 263, 208],
    [86, 307, 20, 44, 308, 340, 313, 251],
    [45, 285, 46, 154, 241, 261, 236, 158],
    [48, 307, 346, 43, 251, 49, 232, 297],
    [55, 343, 13, 59, 350, 342, 361, 111],
    [353, 239, 70, 307, 260, 40, 177, 210],
    [195, 308, 304, 251, 4, 8, 43, 193],
    [135, 136, 308, 304, 43, 44, 262, 251],
    [99, 284, 333, 87, 40, 241, 360, 321],
    [59, 55, 255, 276, 394, 153, 275, 200],
    [262, 236, 18, 55, 135, 256, 16],
    [46, 257, 43, 341, 251, 315, 172, 314],
    [360, 46, 172, 105, 295, 156, 333, 143],
    [128, 46, 191, 50, 349, 58, 299, 219],
    [104, 154, 98, 110, 48, 164, 261, 310],
    [259, 231, 44, 308, 298, 147, 106, 346],
    [107, 38, 108, 185, 46, 314, 143, 48],
    [92, 208, 59, 346, 86, 200, 307, 255],
    [270, 159, 357, 48, 111, 154, 346, 46],
    [84, 307, 87, 196, 285, 304, 274, 199],
    [173, 43, 27, 55, 314, 12, 116, 359],
    [285, 208, 207, 342, 307, 308, 48, 49],
    [98, 120, 40, 162, 99, 43, 251, 55],
    [174, 316, 40, 74, 216, 226, 382, 157],
    [199, 299, 150, 276, 105, 235, 329, 135],
    [341, 330, 69, 241, 29, 252, 313, 8],
    [362, 4, 8, 316, 15, 321, 233, 18],
    [149, 148, 385, 18, 350, 300, 382, 319],
    [281, 285, 236, 43, 258, 251, 239, 84],
    [185, 224, 179, 65, 156, 6, 211, 327],
    [59, 10, 343, 257, 236, 337, 304, 56],
    [247, 29, 342, 39, 129, 49, 48, 76],
    [183, 357, 174, 18, 98, 279, 143, 269],
    [333, 226, 300, 39, 109, 145, 215, 170],
    [5, 65, 15, 359, 245, 210, 162, 278],
    [252, 320, 283, 41, 328, 240, 82, 29],
    [307, 308, 44, 43, 258, 359, 257, 154],
    [314, 175, 40, 374, 165, 29, 333, 100],
    [236, 241, 273, 355, 268, 358],
    [261, 48, 162, 154, 333, 105, 316, 1],
    [229, 314, 237, 108, 70, 117, 40, 97],
    [352, 280, 275, 357, 103, 137, 97, 95],
    [162, 331, 297, 105, 284, 236, 22, 29],
    [276, 19, 352, 74, 48, 261, 254, 111],
    [165, 166, 105, 111, 59, 104, 299, 296],
    [172, 360, 37, 86, 340, 41, 258, 241],
    [49, 285, 123, 45, 308, 73, 52, 313],
    [276, 285, 261, 278, 357, 251, 236, 352],
    [308, 307, 353, 59, 137, 48, 257, 40],
    [251, 259, 55, 53, 308, 44, 43, 111],
    [159, 158, 37, 86, 41, 172, 229, 241],
    [43, 255, 293, 74, 56, 179, 39, 115],
    [18, 148, 150, 258, 135, 256, 232, 382],
    [216, 215, 159, 37, 129, 128, 340],
    [116, 345, 110, 342, 31, 58, 224, 46],
    [105, 104, 116, 115, 297, 132, 157, 57],
    [172, 137, 95, 258, 252, 353, 304, 39],
    [304, 353, 346, 148, 258, 55, 44, 241],
    [353, 16, 350, 18, 252, 10, 136, 283],
]

for proj in projects_without_index:
    inRanked = False
    for student in studentProjectRanks:
        if proj in student:
            inRanked = True
            break
    if inRanked == False:
        project_supervisor_zip = [(i,j) for i,j in project_supervisor_zip if i != proj]

projects, project2supervisor = zip(*project_supervisor_zip)
projects = list(projects)
project2supervisor = list(project2supervisor)

key_projects = {k:v for v, k in enumerate(projects)}

supervisor_key = [project2supervisor.count(v) for v in set(project2supervisor)]

print(len(set(project_supervisors_map)))

'''

num_projects = len(projects)
student_project_rank = []
for student in studentProjectRanks:
    all_proj = [num_projects]*num_projects
    for rank, proj in enumerate(student):
        key = key_projects[proj]
        all_proj[key] = rank+1
    student_project_rank.append(all_proj)
a = np.array(student_project_rank)

np.savetxt('test.txt', a, fmt="%s", delimiter=", ")

supervisors_list = []
for i, x in enumerate(supervisor_key):
    supervisors_list.extend([i+1]*x)
print(len(supervisor_key))
print(supervisors_list)

supervisor_max_zip = [(i,j) for i,j in zip(supervisors, lecturerMax) if i in set(project2supervisor)]
new_sup_max = [j for i,j in supervisor_max_zip]
print(new_sup_max)
print(len(new_sup_max))
# print(lecturerMax)










'''